Add load balancing docs (#2559)
knative/docs: docs/serving/load-balancing/README.md

You can configure load balancing on Knative to place the activator in the request path to act as a load balancer.

This guide explains how you can configure load balancing for your Knative system using the activator.

## Prerequisites
- Ensure that individual pod addressability is enabled.

## Activator pod selection
In general, the system will perform best if the number of existing pods is larger than the number of activators, and those numbers divide equally.

> <!--TODO: Add better documentation about what the activator is; explain the components of load balancing; maybe add a diagram-->

Reviewer: Please replace this TODO with TODO(issue#) for better tracking?

> In general, the system will perform best if the number of existing pods is larger than the number of activators, and those numbers divide equally.
> <!--TODO: Add better documentation about what the activator is; explain the components of load balancing; maybe add a diagram-->
> Knative assigns a subset of activators for each revision, depending on the revision size. More revision pods will mean a greater number of activators for that revision. Activators are scaled horizontally, so there may be multiple activators in a deployment.

Reviewer: ```suggestion
Activators are scaled horizontally, so there may be multiple activators in a deployment. In general, the system will perform best if the number of existing pods is larger than the number of activators, and those numbers divide equally.
<!--TODO: Add better documentation about what the activator is; explain the components of load balancing; maybe add a diagram-->
Knative assigns a subset of activators for each revision, depending on the revision size. More revision pods will mean a greater number of activators for that revision.
```

The activator load balancing algorithm works as follows:
- If concurrency is unlimited, the request is sent to the better of two random choices.

> - If concurrency is set to a very low, limited value, the activator will send the request to the first pod that has capacity. Otherwise, if container concurrency is larger, requests will be balanced in a round robin fashion

Reviewer 2: ```suggestion
- If concurrency is set to a value less or equal than 3, the activator will send the request to the first pod that has capacity. Otherwise, requests will be balanced in a round robin fashion
```

<!--TODO: Still think we need to better explain what's low vs a large value here-->
For more information, see the documentation on [concurrency](../../serving/autoscaling/concurrency.md).

## Configuring target burst capacity

> Target burst capacity is mainly responsible for determining whether the activator is in the request path outside of scale from zero scenarios.

Reviewer: ```suggestion
Target burst capacity is mainly responsible for determining whether the activator is in the request path outside of scale-from-zero scenarios.
```

Target burst capacity can be set [per revision](./target-burst-capacity.md), or globally configured using a combination of the following parameters:

* Setting the targeted concurrency limits for the revision. For more information, see the documentation on [concurrency](../../serving/autoscaling/concurrency.md).
* Setting the target utilization parameters. For more information, see the documentation on [target utilization](../../serving/autoscaling/concurrency.md#target-utilization).

<!-- TODO: Move global vs revision level config information out of autoscaling and up to a higher level doc about configmaps, etc.

> Add section that walks through configuring load balancing globally-->

Reviewer: Please convert TODO to TODO(issue#)? Thanks

Author: Just removed this note because I can't remember the context of why to change it, sorry. Been a while since I looked at this PR. We can follow up another time if it comes up again, but I think we have higher priority stuff to do. Global settings are there now anyway so it's OK.

You can configure load balancing on Knative to place the activator in the request path to act as a load balancer.

This guide explains how you can configure load balancing for your Knative system using the activator.

## Prerequisites

> - Ensure that load balancing using an ingress gateway (for example, Istio or Kourier) is not enabled.

Reviewer 2: We don't disable the ingress (let's remove word gateway, since it's Istio specific) load balancing. It will apply only to distributing requests to activators.

Author: @vagababov still awaiting some clarification on this one?

Author: @vagababov will I just remove this line? I'd like to get this PR closed since it has been months.

- Ensure that individual pod addressability is enabled.

## Activator pod selection
In general, the system will perform best if the number of existing pods is larger than the number of activators, and those numbers divide equally.
<!--TODO: Add better documentation about what the activator is; explain the components of load balancing; maybe add a diagram-->
Knative assigns a subset of activators for each revision, depending on the revision size. More revision pods will mean a greater number of activators for that revision. Activators are scaled horizontally, so there may be multiple activators in a deployment.

The activator load balancing algorithm works as follows:

> - If concurrency is unlimited, the request is sent to a random pod.

Reviewer 2: This is changed to `best of 2 random choices`.

Author: It still doesn't make sense IMHO - a user will want to know what the "2 random choices" refer to. If we mean pods we should just say that, or if we're referring to something else please correct me, but IMHO we should specify where the request is actually sent.

> - If concurrency is set to a limited value, the activator will send the request to the first pod that has capacity.

Reviewer 2: This is now upgraded to
`the request to the first pod that has capacity if container concurrency is very small, otherwise requests will be balanced in a round robin fashion`.

Author: Define very small though? What's the actual cut off concurrency wise before it does round robin?

Author: Added a note since I think we should address this better but I don't want to block the PR any longer on it
